File: packages/dto/src/tables/v2/artworks/artwork.rs
use bdk::prelude::*;

use crate::*;

#[api_model(table = artworks)]
pub struct Artwork {
    #[api_model(primary_key)]
    pub id: i64,

    #[api_model(auto = [insert])]
    pub created_at: i64,
    #[api_model(auto = [insert, update])]
    pub updated_at: i64,

    #[api_model(many_to_one = users)]
    pub owner_id: i64,

    pub title: String,

    #[api_model(nullable)]
    pub description: Option<String>,

    #[api_model(type = JSONB)]
    pub file: File,

    #[api_model(one_to_many = artwork_certifications, foreign_key = artwork_id, aggregator = exist)]
    pub is_certified: bool,

    #[api_model(many_to_many = consensus_votes, foreign_table_name = oracles, foreign_primary_key = oracle_id, foreign_reference_key = consensus_id, aggregator = exist)]
    pub is_voted: bool,

    #[api_model(one_to_many = consensus, foreign_key = artwork_id, aggregator = exist)]
    pub has_consensus: bool,
}

File: ts-packages/web/src/lib/api/models/artwork.ts
import { ConsensusVoteType } from './consensus';
import { FileInfo } from './feeds';

export default interface Artwork {
  id: number;
  created_at: number;
  owner_id: number;
  title: string;
  description?: string;
  file: FileInfo;
  is_certified: boolean;
  is_voted: boolean;
  has_consensus: boolean;
}

export interface ArtworkDetail {
  id: number;
  created_at: number;
  updated_at: number;
  artwork_id: number;
  owner_id: number;
  image: string;
  is_certified: boolean;
}

export interface ArtworkCertificate {
  total_votes: number;
  total_oracles: number;
  approved_votes: number;
  rejected_votes: number;
  certified_at: number;
  voters: {
    nickname: string;
    vote_type: ConsensusVoteType;
    description?: string | null;
  }[];
}
